kernel/kernel-api/src/main/java/io/delta/kernel/internal/SnapshotImpl.java

```diff
@@ -121,4 +125,33 @@ public Path getLogPath() {
     public Path getDataPath() {
         return dataPath;
     }
+
+    /**
+     * Returns the timestamp of the latest commit of this snapshot.
+     * For an uninitialized snapshot, this returns -1.
+     * <p>
+     * When InCommitTimestampTableFeature is enabled, the timestamp
+     * is retrieved from the CommitInfo of the latest commit which
+     * can result in an IO operation.
+     *
+     * @param engine the engine to use for IO operations
+     * @return the timestamp of the latest commit
+     */
+    public long getTimestamp(Engine engine) {
+        if (TableConfig.IN_COMMIT_TIMESTAMPS_ENABLED.fromMetadata(metadata)) {
+            if (!inCommitTimestampOpt.isPresent()) {
+                try {
+                    Optional<CommitInfo> commitInfoOpt = DeltaHistoryManager.getCommitInfoOpt(
+                            engine, logPath, logSegment.version);
+                    inCommitTimestampOpt = Optional.of(CommitInfo.getRequiredInCommitTimestamp(
+                            commitInfoOpt, String.valueOf(logSegment.version)));
+                } catch (IOException e) {
+                    throw new RuntimeException("Failed to get inCommitTimestamp with IO", e);
+                }
+            }
+            return inCommitTimestampOpt.get();
+        } else {
+            return logSegment.lastCommitTimestamp;
```

Comment:
Is the `LogSegment` always initialized with the correct `lastCommitTimestamp`?

Reply:
It's the modification time of the file. When we couldn't get ICT, it's a fallback.